modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/persistence/db/MultipleParallelCacheDeleteDeadlockTest.java

```diff
@@ -0,0 +1,321 @@
+    private static final String CACHE_1 = "cache_1";
+
+    /** */
+    private static final String CACHE_2 = "cache_2";
+
+    /** */
+    private static final String CACHE_GRP_1 = "cache_grp_1";
+
+    /** */
+    private static final String CACHE_GRP_2 = "cache_grp_2";
+
+    /** */
+    private H2TreeIndex.H2TreeFactory regularH2TreeFactory;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setFailureHandler(new StopNodeFailureHandler())
+            .setDataStorageConfiguration(
+                new DataStorageConfiguration().setDefaultDataRegionConfiguration(
+                    new DataRegionConfiguration()
+                        .setPersistenceEnabled(true)
+                        .setInitialSize(10 * 1024L * 1024L)
+                        .setMaxSize(50 * 1024L * 1024L)
+                )
+                .setCheckpointFrequency(Integer.MAX_VALUE)
+            )
+            .setCacheConfiguration(
+                new CacheConfiguration(CACHE_1)
+                    .setGroupName(CACHE_GRP_1)
+                    .setSqlSchema("PUBLIC"),
+                new CacheConfiguration(CACHE_2)
+                    .setGroupName(CACHE_GRP_2)
+                    .setSqlSchema("PUBLIC")
+            );
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        cleanPersistenceDir();
+
+        regularH2TreeFactory = H2TreeIndex.h2TreeFactory;
+
+        H2TreeIndex.h2TreeFactory = H2TreeTest::new;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        H2TreeIndex.h2TreeFactory = regularH2TreeFactory;
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+
+        super.afterTest();
+    }
+
+    /** */
+    @Test
+    public void test() throws Exception {
+        IgniteEx ignite = startGrids(1);
+
+        ignite.cluster().active(true);
+
+        IgniteCache cache1 = ignite.getOrCreateCache(CACHE_1);
+        IgniteCache cache2 = ignite.getOrCreateCache(CACHE_2);
+
+        query(cache1, "create table t1(id integer primary key, f integer) with \"CACHE_GROUP=" + CACHE_GRP_1 + "\"");
+        query(cache1, "create index idx1 on t1(f)");
+
+        for (int i = 0; i < 500; i++)
+            query(cache1, "insert into t1 (id, f) values (?, ?)", i, i);
+
+        query(cache2, "create table t2(id integer primary key, f integer) with \"CACHE_GROUP=" + CACHE_GRP_2 + "\"");
+        query(cache2, "create index idx2 on t2(f)");
+
+        for (int i = 0; i < 500; i++)
+            query(cache2, "insert into t2 (id, f) values (?, ?)", i, i);
+
+        Thread checkpointer = new Thread(() -> {
+            try {
+                checkpointBlockingLatch.await();
+
+                forceCheckpoint();
+
+                testCompletionBlockingLatch.countDown();
```

Comment:
Let's make count down in finally section to allow "destroyCaches" thead to proceed in case of failure.
Otherwise, "destroyCaches" thread can hang forever.
Also it would be polite to interrupt these threads on test failure, and use thread.join() to be sure threads are stopped 
and they will not hold the resources (memory) and will not affect other tests.